File: enc/tools/tool.checklinks.pl
#!/usr/bin/perl -w

#######################################
# проверяет ссылки по всем html-ям
use strict;

my $toolpath = "";
if($^O eq "MSWin32"){
  $toolpath="./tools/";
}

#my @dirs=("../enc_rus/meta","../enc_eng/meta","../enc_rus2/meta");
my @dirs=("../enc_rus3.work/meta");
foreach my $dir (@dirs) {
  srch($dir);
}

#$dir = "../enc_eng2/meta";
#srch($dir);

sub srch
{
 my $dr1 = shift;
 my $dr=$dr1."/";
 #printf "$dr1\n";
 my $ls=$toolpath."ls --ignore=.svn ".$dr1;
 foreach my $fn (`$ls`){
   chomp($fn);
   my $ffn=$dr.$fn;
   if (-d $ffn) {
     srch($ffn);
   } elsif ($fn=~/\.html$/) {
     open my $f, '<', $ffn;
     my @FData = <$f>;
     close $f;
     foreach (@FData) {
       if (/(href|src)=["'] ([^\'\"\#]+?) [\'\"\#]/xi){
         my $ff = $2;
         unless($ff=~/http|mailto|ftp|news/ || -r $dr.$ff){
           if(!($ff =~ /win32\//)){
             print "! $ffn - $ff\n";
           }
         }
       }
     }
   }
 }
}

File: enc/tools/tool.checklinks.py
#!/usr/bin/env python3

"""
Check for broken links
"""

from config import *

from os import walk
from os.path import isfile, join
import re

def check_links(dir_to_check):
  link_match = re.compile(r'(href|src)\s*=[\s"\']*([^#"\']+)', re.I)
  for root, dirs, files in walk(dir_to_check):
    for f in files:
      if (not f.endswith(".html")):
        continue
      infile = open(join(root, f), encoding="utf-8-sig")
      for line in infile:
        for r in link_match.findall(line):
          if (r[1].startswith(("http:", "https:", "mailto:", "ftp:", "news:", "mk:"))):
            continue
          if (not isfile(join(root, r[1]))):
            print("%s - %s" % (join(root, f)[len(dir_to_check)+1:], r[1]))
      infile.close()

check_links(join(ROOT_DIR,"enc_rus","meta"))
#check_links(join(ROOT_DIR,"enc_eng","meta"))
